Mathlib/RingTheory/Binomial.lean

```diff
@@ -0,0 +1,124 @@
+/-
+Copyright (c) 2023 Scott Carnahan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Carnahan
+-/
+
+import Mathlib.Data.Rat.NNRat
+import Mathlib.RingTheory.Polynomial.Pochhammer
+
+/-!
+# Binomial semirings and binomial rings
+
+In this file we introduce binomial semirings and binomial rings, together with related functions
+like binomial coefficients.
+
+According to our main reference [Elliott] (which lists many equivalent conditions), a binomial ring
+is a torsion-free commutative ring R such that for any x ∈ R and any k ∈ ℕ, the product
+x(x-1)⋯(x-k+1) ∈ (k!)R.  The torsion-free condition lets us divide by k! unambiguously, so we get
+uniquely defined binomial coefficients.
+
+The defining condition doesn't require commutativity, and we get a theory with essentially the same
+power by replacing subtraction with addition.  Thus, we consider a semiring R in which
+multiplication by factorials is injective, and demand that for all x ∈ R and any k ∈ ℕ,
+k! | x(x+1)⋯(x+(k-1)).
```

Comment:
Please either use tex and put this (and other pseudo code mixed into comments) in double dollar signs or use backticks ` and lean-ish code

Reply:
Thank you, I think I got all of them.